### ISSUE TITLE
Tensorflow lite GPUDelegate with Android API 31

### ISSUE THREAD

**Issue by amol-tw** [tensorflow/tensorflow#56533]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes 
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**: Pixel 6 pro
-   **TensorFlow installed from (source or binary)**: binary
-   **TensorFlow version (use command below)**: 2.9.0
-   **Python version**:  NA
-   **Bazel version (if compiling from source)**: NA
-   **GCC/Compiler version (if compiling from source)**: NA
-   **CUDA/cuDNN version**: NA
-   **GPU model and memory**: NA
-   **Exact command to reproduce**: NA

### Describe the problem

My model is a classifier trained using EfficientNet and runs on the camera feed using CameraX APIs. It expects a FloatBuffer of size 768 and returns a list of labels. This was working fine upto API 30, but once I update the `targetSdk` to 31 and run on a device with Android 12, I get this crash:

```
    java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
    Falling back to OpenGL
    TfLiteGpuDelegate Init: Batch size mismatch, expected 768 but got 1
    TfLiteGpuDelegate Prepare: delegate is not initialized
    Node number 247 (TfLiteGpuDelegateV2) failed to prepare.
    Restored original execution plan after delegate application failure.
        at org.tensorflow.lite.NativeInterpreterWrapper.createInterpreter(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.init(NativeInterpreterWrapper.java:100)
        at org.tensorflow.lite.NativeInterpreterWrapper.<init>(NativeInterpreterWrapper.java:52)
        at org.tensorflow.lite.NativeInterpreterWrapperExperimental.<init>(NativeInterpreterWrapperExperimental.java:40)
        at org.tensorflow.lite.Interpreter.<init>(Interpreter.java:192)
```

### Source code / logs
Initialization code: 

```
fun initialize() {
        val compatList = CompatibilityList()
        val options = Interpreter.Options().apply {
            if (compatList.isDelegateSupportedOnThisDevice) {
                val delegateOptions = compatList.bestOptionsForThisDevice
                addDelegate(GpuDelegate(delegateOptions))
            }
        }
        modelDownloader.download {
            interpreter = Interpreter(it, options)
        }
    }
```

The crash occurs on the line `interpreter = Interpreter(it, options)` 

Inference code (never gets called in case of crash) : 
```
        val outputs: MutableMap<Int?, Any?> = HashMap()
        val result = Array(1) { FloatArray(8) }
        outputs[0] = result
        val input = passportInputImageConverter(imageProxy.image!!)
        interpreter!!.runForMultipleInputsOutputs(
            arrayOf(input), outputs
        )

```
where `passportInputImageConverter` would crop and convert the image to FloatBuffer using :  

```
 private val imageProcessor: ImageProcessor = ImageProcessor.Builder()
            .add(ResizeOp(INPUT_SIZE, INPUT_SIZE, ResizeOp.ResizeMethod.NEAREST_NEIGHBOR))
            .add(NormalizeOp(0f, 255f))
            .build()
```

Dependency list:

```
org.tensorflow:tensorflow-lite:2.9.0
org.tensorflow:tensorflow-lite-gpu:2.9.0
org.tensorflow:tensorflow-lite-gpu-delegate-plugin:0.4.1
org.tensorflow:tensorflow-lite-support:0.4.1
```

**Comment by mohantym** [comment]
Hi @amol-tw ! Could you share the inference code snippet too. Thank you!

**Comment by amol-tw** [comment]
@mohantym I have updated the issue with the inference code and more details, but I am afraid that it's never called as the crash happens during the initialization.

**Comment by mohantym** [comment]
Hi @sachinprasadhs ! Could you please look at this issue. Thank you!

**Comment by impjdi** [comment]
The error message tells you what's wrong.  Expected batch dimension of 768
but found 1.  The GPU delegate has some restrictions with the batch
dimensions and requires that dimension to be consistent throughout the
network.  In your case, you omitted the batch dimension, so your input is
768 instead of 1x768 which confuses the delegate.  I suggest you
re-architect the model to have an explicit batch size of 1 in every input &
intermediate tensors.

**Comment by amol-tw** [comment]
@impjdi Thanks for the response. 

Were the restrictions to GPUDelegate introduced recently? Can you point to me to some documentation around it?

Asking because this does not explain that the model runs fine when `targetSDK` is set to 30 or lower. I tried a couple of other models also and got errors like the following:

```    java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
    Falling back to OpenGL
    TfLiteGpuDelegate Init: Only identical batch dimension is supported
    TfLiteGpuDelegate Prepare: delegate is not initialized
    Node number 247 (TfLiteGpuDelegateV2) failed to prepare.
```

and 

```    Process: com.example.verificationcamerademo, PID: 980
    java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
    Falling back to OpenGL
    TfLiteGpuDelegate Invoke: GpuDelegate must run on the same thread where it was initialized.
    Node number 10 (TfLiteGpuDelegateV2) failed to invoke. ```


Again, these models seem to run absolutely fine on `targetSDK` 30 but crash on 31.

**Comment by roserg** [comment]
"Can not open OpenCL library on this device"
This is root problem with Android API 31.
Looks like with Android API 30 it was able to open libOpenCL.so and use OpenCL delegate.
In Android API 31 settings were changed probably and now libOpenCL not available for usual users.
As result used OpenGL backend. But the problem is that OpenCL has much more ops covered and features(for example support of batch != 1). In your case OpenCL backend was able to run model and OpenGL can not.
I don't think that in nearest future OpenGL will have similar feature parities to OpenCL backend.
So I explained why you have this problem, but I don't know how to help you.
